test-app/src/main/java/org/readium/r2/testapp/reader/ReaderActivity.kt

```diff
@@ -34,50 +36,46 @@ import org.readium.r2.testapp.outline.OutlineFragment
  */
 open class ReaderActivity : AppCompatActivity() {
 
-    protected lateinit var readerFragment: BaseReaderFragment
     private lateinit var modelFactory: ReaderViewModel.Factory
-    private lateinit var publication: Publication
-
-    lateinit var binding: ActivityReaderBinding
+    private lateinit var model: ReaderViewModel
+    private lateinit var binding: ActivityReaderBinding
+    private lateinit var readerFragment: BaseReaderFragment
 
     override fun onCreate(savedInstanceState: Bundle?) {
-        val inputData = ReaderContract.parseIntent(this)
-        modelFactory = ReaderViewModel.Factory(applicationContext, inputData)
+        val arguments = ReaderActivityContract.parseIntent(this)
+        val app = applicationContext as Application
+        modelFactory = ReaderViewModel.Factory(app, arguments)
+        model = ViewModelProvider(this)[ReaderViewModel::class.java]
+
+        /*
+         * [ReaderViewModel.Factory] provides dummy publications if the [ReaderActivity] is restored
+         * after the app process was killed because the [ReaderRepository] is empty.
+         * In that case, finish the activity as soon as possible and go back to the previous one.
+         */
+        if (model.publication.readingOrder.isEmpty()) {
+            finish()
```

Comment:
Some navigator fragments will crash there if we continue with an empty `readingOrder`.
```suggestion
            finish()
            super.onCreate(savedInstanceState)
            setContentView(View())
            return
```

Reply:
Isn't it exactly the way it used to work with intent helpers?

Reply:
The intent helpers couldn't act at this level, but this was supposed to be the "correct" way to use it in the test app. It's possible that it wasn't implemented properly before.

For example in `PdfNavigatorFragment`, you have this in the `onCreateView`:

```
go(publication.readingOrder.first())
```

Also even with this suggestion I'm not sure this is 100% crash-free with the fragments recovery. This part is really shaky like we discussed, hopefully we can implement a better solution by using maybe Navigation component at some point.

Reply:
I think the intent helper acts here because of this line https://github.com/readium/kotlin-toolkit/blob/adeb8f452fbd48b6ab7e1132264bcce9ab64ee95/test-app/src/main/java/org/readium/r2/testapp/reader/ReaderActivity.kt#L47
This is mostly like the new code. The Activity is told to finish, but it doesn't do it immediately. The rest of the code is executed with a publication with an empty reading order, and an exception might be thrown https://github.com/readium/kotlin-toolkit/blob/adeb8f452fbd48b6ab7e1132264bcce9ab64ee95/test-app/src/main/java/org/readium/r2/testapp/reader/ReaderActivity.kt#L74
Anyway, I (hopefully) made the `ReaderActivity` crash-free yesterday.